PerformanceTests/MotionMark/resources/debug-runner/motionmark.js

```diff
@@ -679,5 +682,44 @@ Utilities.extendObject(window.benchmarkController, {
         sectionsManager.setSectionHeader("test-graph", testName);
         sectionsManager.showSection("test-graph", true);
         this.updateGraphData(testResult, testData, benchmarkRunnerClient.results.options);
+    },
+    detectSystemFrameRate: function()
+    {
+        let last = 0;
+        let average = 0;
+        let count = 0;
+        const outputField = document.querySelector("#frame-rate-detection span");
+        const finish = function()
+        {
+            const commonFrameRates = [15, 30, 45, 60, 90, 120, 144];
+            const distanceFromFrameRates = commonFrameRates.map(rate => {
+                return Math.abs(Math.round(rate - average));
+            });
+            let targetFrameRate = 60;
+            for (let i = 0; i < commonFrameRates.length; i++) {
+                if (distanceFromFrameRates[i] < 10) {
+                    targetFrameRate = commonFrameRates[i];
+                    break;
+                }
+            }
+            document.getElementById("frame-rate-detection").textContent = `Detected system frame rate as ${targetFrameRate} FPS`;
+            document.getElementById("system-frame-rate").value = targetFrameRate;
+            document.getElementById("frame-rate").value = Math.round(targetFrameRate * 5 / 6);
+        }
+        const tick = function(t)
+        {
+            t = t || 0;
+            average -= average / 30;
+            average += 1000. / (t - last) / 30;
```

Comment:
It averages 30 frames of input, rather than just picking the current tick. It also doesn't need the data from the first and last frames.

Reply:
But in the first tick average will be meaningless `average = 1000. / t / 30;`

Also I do not understand why we averages 30 frames of input but loop for 300 frames.

Reply:
> What is the value of average when the tick() function is first called and last = 0?

Did you see where `average` and `last` are initialised to 0?

The incoming timestamp will have a value. This means `average` will evaluate to be a very small value. But that's totally fine because we can't use the value from the first frame anyway. 

> But in the first tick average will be meaningless average = 1000. / t / 30;

Exactly! This is why we average 30 frames and only finish after 300.

> Also I do not understand why we average 30 frames of input but we loop for 300 frames.

Because we don't want to take the values from the ramp-up of rAF. By the time we've seen 300 frames (5 seconds on 60fps, 2.5 seconds on 120fps), the most recent 30 frames give a good average. It's not completely foolproof, but it is better than taking the average of all frames. We then pick the closest "known" rate anyway.

Reply:
(the funny averaging is "exponential moving average" (not actually averaging 30 frames of input -- averaging all inputs but with exponential decay) which I use when I am being lazy, and I think dino stole it from a page of mine, so you can blame me for the weirdness 😆)